compiler/src/iree/compiler/GlobalOptimization/Interfaces/HoistableOpInterface.cpp

```diff
@@ -0,0 +1,189 @@
+    // Detect op that only broadcast input as fusing them makes the new
+    // op cheaper.
+    if (genericOp.getNumParallelLoops() == genericOp.getNumLoops() &&
+        isa<linalg::YieldOp>(genericOp.getBody()->front())) {
+      for (OpOperand *opOperand : genericOp.getDpsInputOperands()) {
+        AffineMap indexingMap = genericOp.getMatchingIndexingMap(opOperand);
+        if (indexingMap.isProjectedPermutation() &&
+            indexingMap.getNumDims() != indexingMap.getNumResults()) {
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+  bool isAtomicallyHoistableOp(Operation *) const { return true; }
+  bool isOperandHoistable(Operation *op, OpOperand *operand) const {
+    linalg::LinalgOp linalgOp = llvm::cast<linalg::LinalgOp>(op);
+    // For linalg ops, we only want to hoist inputs.
+    return operand->getOperandNumber() < linalgOp.getNumDpsInputs();
+  }
+};
+
+//===----------------------------------------------------------------------===//
+// Interface Registration
+//===----------------------------------------------------------------------===//
+
+/// Helper structures that iterates over all Op types in `OpTys` and registers
+/// the associated Hoistable___OpInterface.
+template <typename... Ops>
+struct UnhoistableOpInterfaceHelper {
+  static void registerOpInterface(MLIRContext *ctx) {
+    (Ops::template attachInterface<UnhoistableOpInterface<Ops>>(*ctx), ...);
+  }
+};
+
+template <typename... Ops>
+struct HoistableNonLeafOpInterfaceHelper {
+  static void registerOpInterface(MLIRContext *ctx) {
+    (Ops::template attachInterface<HoistableNonLeafOpInterface<Ops>>(*ctx),
+     ...);
+  }
+};
+
+template <typename... Ops>
+struct AlwaysHoistableOpInterfaceHelper {
+  static void registerOpInterface(MLIRContext *ctx) {
+    (Ops::template attachInterface<AlwaysHoistableOpInterface<Ops>>(*ctx), ...);
+  }
+};
+
+template <typename... Ops>
+struct HoistableFlowOpInterfaceHelper {
+  static void registerOpInterface(MLIRContext *ctx) {
+    (Ops::template attachInterface<HoistableFlowOpInterface<Ops>>(*ctx), ...);
+  }
+};
+
+template <typename... Ops>
+struct HoistableLinalgOpInterfaceHelper {
+  static void registerOpInterface(MLIRContext *ctx) {
+    (Ops::template attachInterface<HoistableLinalgOpInterface<Ops>>(*ctx), ...);
+  }
+};
+
+void registerHoistableOpInterfaces(DialectRegistry &registry) {
+  // Register hoistable type interfaces for Flow ops.
+  registry.addExtension(
+      +[](MLIRContext *ctx, IREE::Flow::FlowDialect *dialect) {
+        // Register all Flow ops. `Flow::StreamableOp` is an interface and it is
+        // not possible to attach an external interface to an existing
+        // interface. Therefore, attach the `HoistableOpInterface` to all ops
+        // one-by-one.
+        HoistableFlowOpInterfaceHelper<
+#define GET_OP_LIST
+#include "iree/compiler/Dialect/Flow/IR/FlowOps.cpp.inc"
+            >::registerOpInterface(ctx);
```

Comment:
I don't think there's actually a way to dynamically register ops based on registered interfaces like I wanted (there might be for traits, but I'm not sure I follow all of the polymorphism magic MLIR does exactly) so I'm just doing some redundant registration of the interface here (i.e for other flow ops that don't really need it). Alternatively I could list out all of the known metadata only ops, but that more or less defeats the purpose of defining the interface based on the metadata status of the op. It's nice that the flow op doesn't have to register this directly.

Reply:
Seems like I can't do this anyway. Bazel doesn't like including a private header.

Reply:
Haven't looked at why you need this but it is allowed if there header is in hdrs, not srcs.

Reply:
Ok well the `.h.inc` doesn't work because I just need to get the (static) list of all linalg structured ops. Seems like that's only exposed with a preprocesser directive `GET_OP_LIST`.